src/drivers/magnetometer/qmc5883/qmc5883.cpp

```diff
@@ -0,0 +1,1640 @@
+	_comms_errors(perf_alloc(PC_COUNT, "qmc5883_com_err")),
+	_range_errors(perf_alloc(PC_COUNT, "qmc5883_rng_err")),
+	_conf_errors(perf_alloc(PC_COUNT, "qmc5883_conf_err")),
+	_sensor_ok(false),
+	_calibrated(false),
+	_rotation(rotation),
+	_range_bits(0),
+	_conf_reg(0),
+	_temperature_counter(0),
+	_temperature_error_count(0)
+{
+	// set the device type from the interface
+	_device_id.devid_s.bus_type = _interface->get_device_bus_type();
+	_device_id.devid_s.bus = _interface->get_device_bus();
+	_device_id.devid_s.address = _interface->get_device_address();
+	_device_id.devid_s.devtype = DRV_MAG_DEVTYPE_QMC5883;
+
+	// enable debug() calls
+	_debug_enabled = false;
+
+	// default scaling
+	_scale.x_offset = 0;
+	_scale.x_scale = 1.0f;
+	_scale.y_offset = 0;
+	_scale.y_scale = 1.0f;
+	_scale.z_offset = 0;
+	_scale.z_scale = 1.0f;
+
+	// work_cancel in the dtor will explode if we don't do this...
+	memset(&_work, 0, sizeof(_work));
+}
+
+QMC5883::~QMC5883()
+{
+	/* make sure we are truly inactive */
+	stop();
+
+	if (_reports != nullptr) {
+		delete _reports;
+	}
+
+	if (_class_instance != -1) {
+		unregister_class_devname(MAG_BASE_DEVICE_PATH, _class_instance);
+	}
+
+	// free perf counters
+	perf_free(_sample_perf);
+	perf_free(_comms_errors);
+	perf_free(_range_errors);
+	perf_free(_conf_errors);
+}
+
+int
+QMC5883::init()
+{
+	int ret = PX4_ERROR;
+
+	ret = CDev::init();
+
+	if (ret != OK) {
+		DEVICE_DEBUG("CDev init failed");
+		goto out;
+	}
+
+	/* allocate basic report buffers */
+	_reports = new ringbuffer::RingBuffer(2, sizeof(mag_report));
+
+	if (_reports == nullptr) {
+		goto out;
+	}
+
+	/* reset the device configuration */
+	reset();
+
+	_class_instance = register_class_devname(MAG_BASE_DEVICE_PATH);
+
+	ret = OK;
+	/* sensor is ok, but not calibrated */
+	_sensor_ok = true;
+out:
+	return ret;
+}
+
+int QMC5883::set_range(unsigned range)
+{
+	int ret;
+	if (range <= 2) {
+		_range_bits = 0x00;
+		_range_scale = 1.2e-4;
+		_range_ga = 2.00f;
+		ret = write_reg(QMC5883_ADDR_CONTROL_1, QMC5883_OUTPUT_RANGE_2G);
+	} else {
+		_range_bits = 0x01;
+		_range_scale = 3.0e-3;
+		_range_ga = 8.00f;
+		ret = write_reg(QMC5883_ADDR_CONTROL_1, QMC5883_OUTPUT_RANGE_8G);
+	}
+
+	if (OK != ret) {
+		perf_count(_comms_errors);
+	}
+
+	uint8_t range_bits_in = 0;
+	ret = read_reg(QMC5883_ADDR_CONTROL_1, range_bits_in);
+
+	if (OK != ret) {
+		perf_count(_comms_errors);
+	}
+
+	return !(range_bits_in == (_range_bits << 4));
+}
+
+/**
+   check that the range register has the right value. This is done
+   periodically to cope with I2C bus noise causing the range of the
+   compass changing.
+ */
+void QMC5883::check_range(void)
+{
+	int ret;
+
+	uint8_t range_bits_in = 0;
+	ret = read_reg(QMC5883_ADDR_CONTROL_1, range_bits_in);
+
+	if (OK != ret) {
+		perf_count(_comms_errors);
+		return;
+	}
+
+	if (range_bits_in != _range_bits << 4) {
+		perf_count(_range_errors);
+		ret = write_reg(QMC5883_ADDR_CONTROL_1, (_range_bits << 4));
```

Comment:
@dagar I think this is why it was going STALE.  The first 2 bits of the control address (0x09) are the data rate which are getting written over here and putting it back into standby mode.